### PR TITLE
Allow configuration of upstream connection limits in Envoy

### LINKED ISSUE #6359: Connect: allow configuration of upstream connection pool limits
**Scope:**
 - No RFC needed
 - Some minor UX work to decide best config to support that will be maximally future compatible with other proxy integrations. The rationale can be given in PR description and signed off as part of review.
 - O(days) task to add the config, plumb though to Envoy and write tests (assuming familiarity)

---

Allow configuration of Envoy's client side connection/request limiting. 

Note this is **NOT** the global rate limit filter that calls an external service which is most of what you find in the envoy docs for rate limiting. 

Envoy somewhat confusingly calls this "Circuit Breakers". (The thing most other systems call a "Circuit breaker" Envoy calls "Outlier Detection"). See https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto.

I propose allowing both `max_connections` and `max_pending_requests`/`max_requests` to be controlled in appropriate services (depends on the protocol being L4 vs L7). 

Bear in mind that the UX we build to expose these needs to be general enough to map to other proxies reasonably well too. The semantics of max_[pending_]requests are not exactly a "rate limit" in terms of req/s but I think HAProxy and Nginx can both support similar semantics to these: e.g. HAProxy maxconn behaves the same as max_requests I think (please check) and maxqueue like max_pending_requests. nginx has max_conns and queue which seem equivalent but are in Commercial version only. Exact identical behaviour is not so important more just that we are not adding things to our first-class config UX that can't work in anything but Envoy or works wildly differently everywhere else etc.

### PR DESCRIPTION
**Pull request by crhino**
Fixes #6359 

This PR adds a `limits` field to the upstream configuration of a connect envoy proxy. This `limits` field is intended to contain configuration regarding connections/requests of the given service to the specified upstream, such as `max_connections`, `max_pending_requests`, and `max_requests`.

This is currently only allowed on the `proxy.upstreams[*].config` part of the service definition, a future improvement would be to allow these upstream configurations to be set in a service's `service-defaults` central config.

Example:
```
services {
  name = "s1"
  port = 8080
  connect {
    sidecar_service {
      proxy {
        upstreams = [
          {
            destination_name = "s2"
            local_bind_port = 5000
            config {
              limits {
                max_connections = 3
                max_pending_requests = 4
                max_requests = 5
              }
            }
          }
        ]
      }
    }
  }
}
```

## UX

Some relevant UX decisions I made, looking for feedback on this.

### Limits Block

I chose to scope this section under a `limits` block in order to more concretely define them as related, as well as simplify the implementation. This should be a natural place to add more related configuration if necessary as well.

I considered naming this something to the effect of `connection_limits`, but felt that was a little too wordy when typing.

### max_* Naming

After looking into the big three proxy configuration (Envoy, Nginx, HAProxy), I eventually settled on keeping the naming that Envoy uses for these values.

For `max_connections` and `max_pending_requests`, all three had similar configuration values, although they varied in whether those values were set on a per-server or per-{cluster,frontend} basis.

For `max_requests`, this represents the max allowed in-flight requests to an upstream cluster by Envoy. This is mostly useful in a HTTP/2 context, since HTTP/1.1 requests are mostly bounded by the max_connections parameter. I was not able to find an equivalent configuration option in HAProxy or Nginx.

I also considered changing `max_requests` to `max_inflight_requests` or `max_concurrent_requests` to be more specific in naming, but decided it was better to keep 100% consistent with Envoy than to change a single name.

## TODO

- [x] Update all related documentation with this new configuration value.
- [x] Decide whether to support `service-defaults` in this PR or future one.

### CODE REVIEW

**Comment by crhino** [comment]
Another high-level question I have regarding this configuration/PR is whether we want to use this opportunity to add support for configuring upstream values like these in `service-defaults` now, or have that be a future improvement.

From my understanding, centralizing default configuration in the `service-defaults` is the way we want to trend towards, and thus only supporting this in the service definition feels like a slight step backwards.

**Comment by banks** [comment]
Great job @crhino!

I've not yet dug through code but will soon but want to share initial feedback on the PR description and questions.

Overall I think this looks great and is a sensible place to start. I think the next discussion after this PR is where the right place to add this in `service-defaults` is (or possible `service-resolver` but that's a longer conversation...) but let's do that separately.

On the naming choosen, I think all of it is great and appreciate the research and rationale there.

My only suggestion would be that `max_requests` alone is not super intuitive - it might seem like a hard limit on the total number of requests (which makes no sense) but then could be a maximum number per conection before cycling or per stream or... What do you think about calling it `max_requests_inflight` or `max_concurrent_requests`?

> whether we want to use this opportunity to add support for configuring upstream values like these in service-defaults now

You're exactly right this is the plan but it's actually way more subtle and difficult that it seems - I have an RFC in progress for it but the sticky part is resolving how upstream ports are assigned and discovered once the definition isn't tied to a specific instance.

So yes that's coming but we shouldn't mix it up here just yet!

**Comment by crhino** [comment]
I agree `max_concurrent_requests` is more descriptive than `max_requests`. My only hesitance is around whether to adhere strictly to Envoy naming conventions, or change to be more descriptive.

I don't think I have a strong opinion on this naming, other than I like `max_concurrent_requests` more than `max_requests_inflight`. Will update names according and see how it feels.

File: agent/xds/clusters.go [hunk 1]
```diff
@@ -449,3 +455,27 @@ func (s *Server) makeMeshGatewayCluster(clusterName string, cfgSnap *proxycfg.Co
...
+	// Make sure to not create any thresholds when passed the zero-value in order
+	// to rely on Envoy defaults
+	if limits == empty {
+		return nil
+	}
+
+	threshold := &envoycluster.CircuitBreakers_Thresholds{}
+	// Likewise, make sure to not set any threshold values on the zero-value in
+	// order to rely on Envoy defaults
+	if limits.MaxConnections != 0 {
```
Comment (banks) [inline agent/xds/clusters.go]: I wonder if it's ever valid in Envoy to want to explicitly set these to `0` to disable them or similar? If so we'd need something a little more sophisticated to be able to disambiguate "not set" and "explicit zero"...

At any rate it's likely an edge case but perhaps `-1` could be purposed for this?

**Comment by crhino** [comment]
[Link to the preview docs website](https://5ddd45809897e05fe078fa1f--consul-docs-preview.netlify.com/docs/connect/proxies/envoy.html#proxy-upstream-config-options)

File: agent/xds/clusters.go (hunk 1 above)
Comment (crhino) [reply agent/xds/clusters.go]: An interesting thought experiment around why someone might want to set that, but I agree we should not make it impossible for someone to do that.

Instead of `-1` as a sentinel value, I am slightly inclined to use `*int` types and represent `nil` as the zero value. Not sure if we have any standards around this elsewhere.

Comment (banks) [reply agent/xds/clusters.go]: We use pointers in the config package so we can internally distinguish between not set and zero, but we don't have that anywhere in public API currently. I'm not quite sure how it would work in mapstructure although it probably would I guess and it would therefore be "invisible" to users which is nice so +1.

Where it does end up being visible, we have at least some precedent for using `-1` for disable in our `ports{}` config. In that case it wasn't because of the default zero thing since config already uses pointers, it was because port `0` is actually a valid thing too (meaning let the OS pick a free one for me).

Honestly, I'm on the fence about whether it's even worth supporting yet especially given these specific parameters - setting a really high number like 1000000 is also a fine work around if you really don't want to choose a sensible limit anyway. But yeah pointer would be cleaner in this case.

Comment (crhino) [reply agent/xds/clusters.go]: Yes, I think using pointer values is pretty clean and sets a nice precedent for any additional properties that are added.

Also, to be clear, the Envoy default values if something is not set is actually `1024` for all three configuration values, not unlimited. So setting something really high like 1000000 is actually an explicit choice away from the default.

Comment (banks) [reply agent/xds/clusters.go]: Does Envoy discriminate between explicit `0` and unset? If no this is a moot conversation anyway!

But glad we came up with a clean solution in general 😄?

Comment (crhino) [reply agent/xds/clusters.go]: Good question. Did a quick test and envoy does let you set `max_connections: 0` and then not allow any connections to be made to the upstream. 👍
